### DISCUSSION TITLE
Problems with multiple remotes

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#3871] [Q&A]
I was trying to update [a branch in my fork](https://github.com/martinvonz/jj/pull/3780) to resolve conflicts. After rebasing onto `main@origin` and resolving the conflict, I used `jj git push`, which...created a branch in `martinvonz/jj` rather than updating the remote from which I was tracking the branch.

As far as I know, `jj branch` doesn't have any way to delete a branch remotely without deleting it locally, or to specify only a particular remote on which to delete the branch. So I simply used `git push --delete origin descendants-range`.

But that caused `jj` to try to delete the branch anyway (note that the branch name is not present in the `status` output):

```
$> jj branch track descendants-range@personal
Done importing changes from the underlying Git repo.
Warning: Remote branch already tracked: descendants-range@personal
Nothing changed.
$> jj status
The working copy is clean
Working copy : prqmxzzk 9d9d8057 (empty) (no description set)
Parent commit: zswoqyzt 66989667 give descendants a range arg
$> jj branch set descendants-range -r @-
Error: No such branch: descendants-range
Hint: Use `jj branch create` to create it.
$> jj branch list
descendants-range (deleted)
  @personal: zswoqyzt hidden 3c1a7c7a give descendants a range arg
main: svposxsm e3bb825a jj git push: remove the NotFastForward error
Hint: Branches marked as deleted will be *deleted permanently* on the remote on the next `jj git push`. Use `jj branch forget` to prevent this.
```

This feels similar to #2504 in that it's a surprising (to me) change to the `jj` state that is automatically and unavoidably applied based on a change in the `git` state. But in this case, I'm frankly kind of baffled by the behavior:

* The branch was created via `jj git branch track <name>@personal`. Why did `jj git push` decide to push to `origin` rather than `personal`? I rebased onto `main@origin` rather than tracking `main@origin` and rebasing onto (local) `main`, but I don't see why that would affect which remote is used for tracking.
* `git push --delete` does *not* delete the local branch. So why did importing the git changes delete the local `jj` branch?
* ...once the `jj` branch was deleted, it was...very hard to recover. I understand why `branch track` is a no-op when the branch is marked for deletion (though the warning message should probably mention that), but there doesn't seem to be any good way to un-mark a branch for deletion. `jj git fetch` simply gave me `Nothing changed` (which seems weird because I actually started with `personal` as the only remote and added `origin` later, so the `fetch` default must be based on the *name* `origin` itself?), so I had to explicitly do `jj branch forget descendants-range; jj git fetch --remote personal; jj branch track descendants-range@personal`.

**Comment by BatmanAoD** [comment]
...come to think of it, I don't actually see how to update the `main` branch in my `personal` remote:

```
$> jj git push -b main --remote personal
Error: Non-tracking remote branch main@personal exists
Hint: Run `jj branch track main@personal` to import the remote branch.
```

Well, of course I don't want to track `main@personal`, I want to move it!

**Comment by yuja** [comment]
> * The branch was created via `jj git branch track <name>@personal`. Why did `jj git push` decide to push to `origin` rather than `personal`? I rebased onto `main@origin` rather than tracking `main@origin` and rebasing onto (local) `main`, but I don't see why that would affect which remote is used for tracking.

Unlike Git, jj's local branch ~has~ doesn't have a notion of single upstream remote. The remote to push is purely selected by `jj git push` arguments (or config option.)

https://github.com/martinvonz/jj/issues/1278

> * `git push --delete` does _not_ delete the local branch. So why did importing the git changes delete the local `jj` branch?

`git push --delete` basically created state after pulling deleted branch from remote, and the deleted remote branch was then imported because it was still tracked.

> * ...once the `jj` branch was deleted, it was...very hard to recover. I understand why `branch track` is a no-op when the branch is marked for deletion (though the warning message should probably mention that), but there doesn't seem to be any good way to un-mark a branch for deletion. `jj git fetch` simply gave me `Nothing changed` (which seems weird because I actually started with `personal` as the only remote and added `origin` later, so the `fetch` default must be based on the _name_ `origin` itself?), so I had to explicitly do `jj branch forget descendants-range; jj git fetch --remote personal; jj branch track descendants-range@personal`.

I think you can use `jj branch create` to revive the local branch.
Another easier option is to `jj branch forget` and re-fetch remote branches from all remotes.

**Reply by BatmanAoD** [reply]
Ah, I hadn't yet seen that the `fetch` and `push` remotes are configurable; that makes this easier. And yes, #1278 looks like it should solve the issue overall.

> `git push --delete` basically created state after pulling deleted branch from remote, and the deleted remote branch was then imported because it was still tracked.

I guess that makes sense -- it's as if the remote branch were deleted and I did a `fetch`.

> I think you can use jj branch create to revive the local branch.

Will `jj branch create` automatically track the branch? Or is this a special case where the "deleted" branch is already tracked, so `branch create` will just "adopt" the tracking info?

> Another easier option is to jj branch forget and re-fetch remote branches from all remotes.

That's effectively what I ended up needing to do.

**Reply by yuja** [reply]
> Will `jj branch create` automatically track the branch? Or is this a special case where the "deleted" branch is already tracked, so `branch create` will just "adopt" the tracking info?

`jj branch create` will keep the existing tracking state, but it seems a bit odd from UX. A deleted branch is still listed as "(deleted)", so it's not completely new to the user. It might be better to allow `jj branch set` instead.

Implementation-wise, tracking state is a property of remote branch, not local branch, and `jj branch delete`/`set`/`create` only touches local branch state.
